ipwb/util.py

```diff
@@ -212,10 +226,11 @@ def readIPFSConfig():
     try:
         with open(ipfsConfigPath, 'r') as f:
             return json.load(f)
-    except IOError:
-        logError("IPFS config not found.")
-        logError("Have you installed ipfs and run ipfs init?")
-        sys.exit()
+
+    except IOError as err:
+        raise Exception(
+            'IPFS config not found. Have you installed ipfs and run ipfs init?'
+        ) from err
```

Comment:
Since `sys.exit()` call is removed, and the wrapper exception handler will simply log the exception message when DEBUG is set to false, how will the process terminate? There is no point in keeping the service alive in an defined state when critical pieces are missing, no matter DEBUG is on or off. DEBUG should make logging more verbose with stack trace, but it should not change the process termination behavior. Current implementation, if I could understand it well, means it will terminate the process by raising the exception when DEBUG is on even when the exception is rather innocent and handlable and will not terminate the process even on critical exceptions when the DEBUG is off. Is this the case or am I missing something here?

Reply:
@ibnesayeed let us describe the calls sequence:

```python
if __name__ == "__main__":
    main()
```

seemingly calls `main()`, but _actually_ calls `wrapper(*args, **kwargs)` (see the implementation of `exception_logger`).

Then, `wrapper` calls the original `main()` inside a `try - except` block.

Now what happens if an unhandled exception pops out somewhere in `main()` or one of the functions it calls? Being unhandled, that exception pops up through the call stack, reaches `main()`, passes it and arrives at `wrapper()`.

The execution of `main()` at this point is already interrupted by a propagating exception. `wrapper()` now can either reraise the exception or just log it - that does not matter because, when its `except` clause is done its work, `wrapper()` function returns and the process stops.

At least it should if everything works as I believe it does.

Reply:
@anatoly-scherbakov I understand how wrapper/decorator functions work in Python. I have actually used them heavily in [an HTTP testing framework](https://github.com/ibnesayeed/webserver-tester) I wrote a while ago. My concern here is a little different. Not every exception needs to be reraised to bubble out of the main and let the process die. For example, if the system is configured to use a local IPFS node (which is the case for now as the code needs some changes to decouple IPFS from IPWB) and the IPFS daemon is not accessible because the binary is not installed on the system then the process should die because there is no hope it will ever be accessible. However, if the daemon is not accessible because it is not running, but is installed, then we may not want to let the process die, because the admin control panel has an option to attempt to start/stop IPFS daemon. In the earlier code, some exceptions were logging a message and returning `None` while other exceptions were logging a message and exiting. I am not sure if that behavior is still preserved. I think it is great that we have a catch-all wrapper exception handler, but we may want to handle some exceptions inline and take appropriate actions instead of bubbling every exception up with a custom message.

Reply:
Never mind, I think the custom exception `IPFSDaemonNotAvailable` is handled inline in the replay. This looks good to me now. Thanks @anatoly-scherbakov for your valuable contributions. We look forward to see more from you.